### PR TITLE
eval: clean up usage of UnwrapDatum

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit removes many calls to `eval.UnwrapDatum` where the first argument was `nil`. In such a case that function is equivalent to `tree.UnwrapDOidWrapper`, so this commit uses the latter wherever possible.

Next, this commit adds an explicit `context.Context` argument to the signature to avoid the usage of the deprecated stored context from `eval.Context`. This required a little bit of plumbing around the index encoding methods.

Release note: None

Epic: None

### CODE REVIEW

File: pkg/sql/importer/exportparquet_test.go [hunk 1]
```diff
@@ -147,11 +146,10 @@ func validateParquetFile(
 				return err
 			}
 
-			// If we're encoding a DOidWrapper, then we want to cast the wrapped datum.
-			// Note that we pass in nil as the first argument since we're not interested
-			// in evaluating the placeholders.
-			validateDatum(t, eval.UnwrapDatum(nil, test.datums[i][j]), eval.UnwrapDatum(nil, datum),
-				test.cols[j].Typ)
+			// If we're encoding a DOidWrapper, then we want to cast the wrapped
+			// datum. Note that we pass we don't use eval.UnwrapDatum since
```
Comment (cucaroach) [inline pkg/sql/importer/exportparquet_test.go:150]: s/"we pass we don't"/"we don't"/

**Review by cucaroach — DISMISSED** [review]
> Reviewed 29 of 29 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89630)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msirek and @yuzefovich)

___
*[`pkg/sql/importer/exportparquet.go` line 779 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89630#-NE505qM27pAFat4Ugj2:-NE505qM27pAFat4Ugj3:btvpqdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f589128003fe49d5d2d93076facde62816b640/pkg/sql/importer/exportparquet.go#L779)):*
> ```Go
> 
> 						// If we're encoding a DOidWrapper, then we want to cast
> 						// the wrapped datum. Note that we pass we don't use
> ```

s/"we pass we don't"/"we don't"/
___
*[`pkg/sql/stats/quantile.go` line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89630#-NE5-l8zAeEiLIXNnM4p:-NE5-l8zAeEiLIXNnM4q:bb5wp9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f589128003fe49d5d2d93076facde62816b640/pkg/sql/stats/quantile.go#L255)):*
> ```Go
> 
> 	var compareCtx eval.Context
> 	compareCtx.SetDeprecatedContext(ctx)
> ```

This seems like a more material change than what's described in the commit notes, why doesn't nil work anymore?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89630)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @msirek)

___
*[`pkg/sql/importer/exportparquet.go` line 779 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89630#-NE505qM27pAFat4Ugj2:-NEZmXsO9tKVxoVWS70a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f589128003fe49d5d2d93076facde62816b640/pkg/sql/importer/exportparquet.go#L779)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

s/"we pass we don't"/"we don't"/
</blockquote></details>

Done.
___
*[`pkg/sql/stats/quantile.go` line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89630#-NE5-l8zAeEiLIXNnM4p:-NEZn0xR0hVr_AEbHoX9:bs7fnup) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f589128003fe49d5d2d93076facde62816b640/pkg/sql/stats/quantile.go#L255)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

This seems like a more material change than what's described in the commit notes, why doesn't nil work anymore?
</blockquote></details>

This is needed because the modified code now explicitly has `eval.Context.deprecatedContext` access in `eval.Context.UnwrapDatum`, so with a `nil` we get a NPE.
___
*[`pkg/sql/importer/exportparquet_test.go` line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89630#-NE-ojqPQ15XAWVWo9uZ-r1-150:-NEZnG3aBwj47vZO9ctq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f589128003fe49d5d2d93076facde62816b640/pkg/sql/importer/exportparquet_test.go#L150)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

s/"we pass we don't"/"we don't"/
</blockquote></details>

Done.

**Review by msirek — APPROVED** [review]
Looks good!

> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89630#-:-NEaiUqB1GOFNTN7D-Wg:b-vs1prh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach)

**Review by michae2 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89630)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach)

___
*[`pkg/sql/stats/quantile.go` line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89630#-NE5-l8zAeEiLIXNnM4p:-NEaq-Wt1RS-9xuangv2:b-v83po7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f589128003fe49d5d2d93076facde62816b640/pkg/sql/stats/quantile.go#L255)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This is needed because the modified code now explicitly has `eval.Context.deprecatedContext` access in `eval.Context.UnwrapDatum`, so with a `nil` we get a NPE.
</blockquote></details>

Would you be willing to change `eval.Context.UnwrapDatum` to still handle a nil `*eval.Context` receiver, as it did before? E.g. maybe something like:

```go
// UnwrapDatum encapsulates UnwrapDatum for use in the tree.CompareContext.
func (ec *Context) UnwrapDatum(d tree.Datum) tree.Datum {
	if ec == nil {
		return UnwrapDatum(nil, nil, d)
	}
	return UnwrapDatum(ec.deprecatedContext, ec, d)
}
```

Before this PR, it was possible to call `Datum.Compare` and `Datum.CompareError` with nil `CompareContext` if the datums don't involve session settings or placeholders, and it would be nice to retain this ability.

It seems silly to have to allocate an `eval.Context` just to not use it.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89630)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @msirek)

___
*[`pkg/sql/stats/quantile.go` line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89630#-NE5-l8zAeEiLIXNnM4p:-NFIAf7dCBj6JspFP3qH:b7xnqxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f589128003fe49d5d2d93076facde62816b640/pkg/sql/stats/quantile.go#L255)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Would you be willing to change `eval.Context.UnwrapDatum` to still handle a nil `*eval.Context` receiver, as it did before? E.g. maybe something like:

```go
// UnwrapDatum encapsulates UnwrapDatum for use in the tree.CompareContext.
func (ec *Context) UnwrapDatum(d tree.Datum) tree.Datum {
	if ec == nil {
		return UnwrapDatum(nil, nil, d)
	}
	return UnwrapDatum(ec.deprecatedContext, ec, d)
}
```

Before this PR, it was possible to call `Datum.Compare` and `Datum.CompareError` with nil `CompareContext` if the datums don't involve session settings or placeholders, and it would be nice to retain this ability.

It seems silly to have to allocate an `eval.Context` just to not use it.
</blockquote></details>

Fair point, refactored.

**Comment by yuzefovich** [comment]
TFTRs!

bors r+

**Pull request merged**